materialize-bigquery/binding.go

```diff
@@ -1,225 +1,319 @@
+	}
 
-	b.store.tempTableName = fmt.Sprintf("%s_store_%d", tempTableNamePrefix, bindingPos)
+	for _, key := range bindingSpec.FieldSelection.AllFields() {
+		columns = append(columns, key)
+		rColumns = append(rColumns, fmt.Sprintf("r.%s", key))
+	}
 
-	if spec.DeltaUpdates {
-		// Store everything (do not merge/upsert).
-		b.store.sql = fmt.Sprintf(`
+	if bindingSpec.DeltaUpdates {
+		loadQuery = "ASSERT false AS 'Load queries should never be executed in Delta Updates mode.';"
+		writeQuery = fmt.Sprintf(`
 		INSERT INTO %s (%s)
 		SELECT %s FROM %s
 		;`,
-			tableDef.Identifier,
-			strings.Join(colIdentifiers, ", "),
-			strings.Join(colIdentifiers, ", "),
-			b.store.tempTableName,
+			br.Table,
+			strings.Join(columns, ", "),
+			strings.Join(rColumns, ", "),
+			br.Table,
 		)
+
 	} else {
-		// Updates for all columns minus the primary keys plus the document field.
-		var lrUpdates []string
-		for _, colName := range append(spec.FieldSelection.Values, spec.FieldSelection.Document) {
-			var col = tableDef.GetColumn(colName)
-			lrUpdates = append(lrUpdates, fmt.Sprintf("l.%s = r.%s", col.Identifier, col.Identifier))
+		loadQuery = fmt.Sprintf(`
+			SELECT %s
+			FROM %s
+			WHERE %s
+			;`,
+			bindingSpec.FieldSelection.Document,
+			br.Table,
+			strings.Join(loadPredicates, " AND "),
+		)
+
+		var joinPredicates []string
+		for _, key := range bindingSpec.FieldSelection.Keys {
+			joinPredicates = append(joinPredicates, fmt.Sprintf("l.%s = r.%s", key, key))
 		}
 
-		// Perform merge query to update existing values.
-		b.store.sql = fmt.Sprintf(`
-		MERGE INTO %s AS l
-		USING %s AS r
-		ON %s
-		WHEN MATCHED AND r.%s IS NULL THEN
-			DELETE
-		WHEN MATCHED THEN
-			UPDATE SET %s
-		WHEN NOT MATCHED THEN
-			INSERT (%s)
-			VALUES (%s)
-		;`,
-			tableDef.Identifier,
-			b.store.tempTableName,
-			strings.Join(pkJoins, " AND "),
-			tableDef.GetColumn(spec.FieldSelection.Document).Identifier,
-			strings.Join(lrUpdates, ", "),
-			strings.Join(colIdentifiers, ", "),
-			strings.Join(rColIdentifiers, ", "),
+		var updates []string
+		for _, k := range append(bindingSpec.FieldSelection.Values, bindingSpec.FieldSelection.Document) {
+			updates = append(updates, fmt.Sprintf("l.%s = r.%s", k, k))
+		}
+
+		writeQuery = fmt.Sprintf(`
+			MERGE INTO %s AS l
+			USING %s as r
+			ON %s
+			WHEN MATCHED AND r.%s IS NULL THEN
+				DELETE
+			WHEN MATCHED THEN
+				UPDATE SET %s
+			WHEN NOT MATCHED THEN
+				INSERT (%s)
+				VALUES (%s)
+			;`,
+			br.Table,
+			externalTableAlias,
+			strings.Join(joinPredicates, " AND "),
+			bindingSpec.FieldSelection.Document,
+			strings.Join(updates, ", "),
+			strings.Join(columns, ", "),
+			strings.Join(rColumns, ", "),
 		)
 	}
+	return &sqlQueries{
+		LoadSQL:  loadQuery,
+		WriteSQL: writeQuery,
+	}
+}
+
+// Bigquery only allows underscore, letters, numbers, and sometimes hyphens for identifiers. Convert everything else to underscore.
```

Comment:
`identifierSanitizer` seems to no longer be used, unless I'm missing it? But also, I think it's probably better to quote such naught identifiers rather than transforming them, anyway. Otherwise, it's easy to end up with duplicate column names.

Reply:
Yes, I meant to use it, and forgot to add it back.